app/javascript/hooks/use_list.js

```diff
@@ -0,0 +1,62 @@
+import { useReducer, useCallback } from 'react'
+
+function reducer(state, action) {
+  switch (action.type) {
+    case 'page.changed':
+      return { ...state, query: { ...state.query, page: action.payload.page } }
+    case 'query.added':
+      return { ...state, query: { ...state.query, ...action.payload.query } }
+    case 'filter.changed':
+      return {
+        ...state,
+        query: { ...state.query, filter: action.payload.filter, page: 1 },
+      }
+    case 'sort.changed':
+      return {
+        ...state,
+        query: { ...state.query, sort: action.payload.sort },
+      }
+    default:
+      if (process.env.NODE_ENV === 'development') {
+        throw new Error(`Unknown action type: ${action.type}`)
+      }
+      return state
+  }
+}
+
+export function useList(initialRequest) {
```

Comment:
To me, it wasn't immediately obvious what `useList` does when one includes it. That said, I t tried coming up with a more descriptive name for this hook but failed.

Reply:
It's so hard!